### PR TITLE
bpo-38962: Fix reference leak in the per-subinterpreter gc

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue38962

Automerge-Triggered-By: @pablogsal

### CODE REVIEW

**Comment by pablogsal** [comment]
I'm not super convinced about this PR, but indeed it fixes the reference leaks in test__xxsubinterpreters (when all other PRs in https://bugs.python.org/issue38962 are applied first):

❯ ./python -m test test__xxsubinterpreters -R :
0:00:00 load avg: 1.10 Run tests sequentially
0:00:00 load avg: 1.10 [1/1] test__xxsubinterpreters
beginning 9 repetitions
123456789
.........

== Tests result: SUCCESS ==

1 test OK.

Total duration: 17.6 sec
Tests result: SUCCESS

File: Python/pylifecycle.c
Comment (vstinner) [inline Python/pylifecycle.c]: I'm not fully comfortable with triggering a GC collection after PyInterpreterState_Clear(). If a finalizer tries to access an interpreter state or a Python thread state, it will likely get an inconsistent object or an error. I'm not sure that finalizers handle well every possible error cases. For example, what happens if a finalizer tries to import a module at this point?

May we could try to split PyInterpreterState_Clear() in two functions or more. Or we could modify _PyGC_CollectNoFail() to not call finalizers. But what's the point of a GC collection in this case?

Maybe I'm too pessimistic and it will be fine.

I suggest to not hold this PR. Subinterpreters is a work in progress anyway.

**Maybe a temporary solution is to not call _PyGC_CollectNoFail() in the main interpreter to mimick Python 3.8 (to reduce the risk of regression), but only call it in subinterpreters.**

**Comment by pablogsal** [comment]
@vstinner  I have modified the PR to trigger the collection only on sub interpreters. Can you check it out?

File: Python/pylifecycle.c
Comment (vstinner) [inline Python/pylifecycle.c]: Does the test also pass if you also put this workaround  in "if (!is_main_interp) {"?

Comment (pablogsal) [reply Python/pylifecycle.c]: Yup, they pass with no leaks if we move the 3 `Py_CLEAR` into the branch.

**Review by vstinner — APPROVED** [review]
LGTM.

As I wrote, I'm not fully satisfied by calling _PyGC_CollectNoFail() after PyInterpreterState_Clear() because of the risk of bugs in objects finalizers, but we cannot fix all Python finalization issues at once. Let's unblock the buildbots, and only later investigate for a better design.

**Pull request merged**
